src/librustc_lint/builtin.rs

```diff
@@ -2153,123 +2153,208 @@ impl ClashingExternDeclarations {
         b: Ty<'tcx>,
         ckind: CItemKind,
     ) -> bool {
-        debug!("structurally_same_type(cx, a = {:?}, b = {:?})", a, b);
-        let tcx = cx.tcx;
-        if a == b || rustc_middle::ty::TyS::same_type(a, b) {
-            // All nominally-same types are structurally same, too.
-            true
-        } else {
-            // Do a full, depth-first comparison between the two.
-            use rustc_middle::ty::TyKind::*;
-            let a_kind = &a.kind;
-            let b_kind = &b.kind;
-
-            let compare_layouts = |a, b| -> bool {
-                let a_layout = &cx.layout_of(a).unwrap().layout.abi;
-                let b_layout = &cx.layout_of(b).unwrap().layout.abi;
-                debug!("{:?} == {:?} = {}", a_layout, b_layout, a_layout == b_layout);
-                a_layout == b_layout
-            };
+        // In order to avoid endlessly recursing on recursive types, we maintain a "seen" set.
+        // We'll need to store every combination of types we encounter anyway, so we also memoize
+        // the result.
+        struct SeenSet<'tcx>(FxHashMap<(Ty<'tcx>, Ty<'tcx>), Option<bool>>);
+
+        enum SeenSetResult {
+            /// We've never seen this combination of types.
+            Unseen,
+            /// We've seen this combination of types, but are still computing the result.
+            Computing,
+            /// We've seen this combination of types, and have already computed the result.
+            Computed(bool),
```

Comment:
Do we need a `HashMap` here? Afaict a hashset would be enough.

`Computed(true)` does the same as `Computing` and `Computed(false)` should have already caused an early exit.

I think it's enough to only store if we already visited a type.

Reply:
(at this point using a new struct here might even be slightly too complex, so we could "just" use `FxHashSet<(Ty<'tcx>, Ty<'tcx>)>` directly here)

Reply:
HashSet would definitely suffice. I've only gone with a map here because I figured if I'm needing to store every combination of types encountered anyway, I might as well also store the result once I know it. 

To be fair, though, now that I think about it, it's probably unlikely that this cache gets a hit within a single invocation of structurally_same_type... Yeah, I'll get rid of it.